File: src/PlayerManager.ts
/**
 * PlayerManager - Manages player setup for Skull King game
 * Handles adding, removing, and validating players before game begins
 */

export interface Player {
  id: string;
  name: string;
  createdAt: number;
}

export class PlayerManager {
  private players: Map<string, Player> = new Map();
  private playerNameSet: Set<string> = new Set();
  private nextId: number = 1;

  private readonly MIN_PLAYERS = 2;
  private readonly MAX_PLAYERS = 8;

  /**
   * Add a new player with validation
   * @param name - Player name (must be unique and non-empty)
   * @returns Player object if successful
   * @throws Error if validation fails
   */
  addPlayer(name: string): Player {
    this.validatePlayerName(name);
    this.validatePlayerCount();

    const player: Player = {
      id: `player_${this.nextId++}`,
      name: name.trim(),
      createdAt: Date.now(),
    };

    this.players.set(player.id, player);
    this.playerNameSet.add(player.name.toLowerCase());

    return player;
  }

  /**
   * Remove a player by ID
   * @param playerId - ID of player to remove
   * @throws Error if player not found
   */
  removePlayer(playerId: string): void {
    const player = this.players.get(playerId);
    if (!player) {
      throw new Error(`Player with ID ${playerId} not found`);
    }

    this.players.delete(playerId);
    this.playerNameSet.delete(player.name.toLowerCase());
  }

  /**
   * Get all current players
   */
  getPlayers(): Player[] {
    return Array.from(this.players.values()).sort(
      (a, b) => a.createdAt - b.createdAt
    );
  }

  /**
   * Get player count
   */
  getPlayerCount(): number {
    return this.players.size;
  }

  /**
   * Check if game can start (2-8 players required)
   */
  canStartGame(): boolean {
    return this.players.size >= this.MIN_PLAYERS && this.players.size <= this.MAX_PLAYERS;
  }

  /**
   * Reset all players (for new game)
   */
  reset(): void {
    this.players.clear();
    this.playerNameSet.clear();
    this.nextId = 1;
  }

  /**
   * Validate player name
   * @throws Error if name is invalid or duplicate
   */
  private validatePlayerName(name: string): void {
    if (!name || typeof name !== 'string') {
      throw new Error('Player name must be a non-empty string');
    }

    const trimmedName = name.trim();
    if (trimmedName.length === 0) {
      throw new Error('Player name cannot be empty or whitespace only');
    }

    if (trimmedName.length > 50) {
      throw new Error('Player name must be 50 characters or less');
    }

    const nameLowercase = trimmedName.toLowerCase();
    if (this.playerNameSet.has(nameLowercase)) {
      throw new Error(
        `Player name "${trimmedName}" is already taken. Player names must be unique.`
      );
    }
  }

  /**
   * Validate player count doesn't exceed maximum
   * @throws Error if at max capacity
   */
  private validatePlayerCount(): void {
    if (this.players.size >= this.MAX_PLAYERS) {
      throw new Error(
        `Maximum of ${this.MAX_PLAYERS} players reached. Cannot add more players.`
      );
    }
  }
}

File: game/player_manager.py
"""Player management module for Skull King game.

Handles player setup, validation, and management during game initialization.
"""


class PlayerManager:
    """Manages players during game setup and play.
    
    Enforces constraints:
    - Player count between 2-8
    - Unique player names (case-insensitive)
    - Player removal during setup
    """
    
    MIN_PLAYERS = 2
    MAX_PLAYERS = 8
    
    def __init__(self):
        """Initialize empty player list."""
        self._players = []
    
    def add_player(self, name):
        """Add a player with validation.
        
        Args:
            name: Player name (string)
            
        Raises:
            ValueError: If name is invalid, duplicate, or player limit reached
            TypeError: If name is not a string
        """
        # Type validation
        if not isinstance(name, str):
            raise TypeError(f"Player name must be a string, got {type(name).__name__}")
        
        # Empty/whitespace validation
        if not name or not name.strip():
            raise ValueError("Player name cannot be empty")
        
        # Normalize the name for storage and checking
        normalized_name = name.strip()
        
        # Check for duplicates (case-insensitive)
        existing_names = [p.lower() for p in self._players]
        if normalized_name.lower() in existing_names:
            raise ValueError(f"Player '{normalized_name}' already exists")
        
        # Check player limit
        if len(self._players) >= self.MAX_PLAYERS:
            raise ValueError(f"Cannot add more than {self.MAX_PLAYERS} players")
        
        # Add the player
        self._players.append(normalized_name)
    
    def remove_player(self, name):
        """Remove a player by name (case-insensitive).
        
        Args:
            name: Player name to remove
            
        Raises:
            ValueError: If player not found
        """
        if not isinstance(name, str):
            raise TypeError(f"Player name must be a string, got {type(name).__name__}")
        
        # Find and remove (case-insensitive)
        for i, player in enumerate(self._players):
            if player.lower() == name.lower():
                self._players.pop(i)
                return
        
        raise ValueError(f"Player '{name}' not found")
    
    def get_players(self):
        """Get list of current players.
        
        Returns:
            List of player names
        """
        return self._players.copy()
    
    def player_count(self):
        """Get current number of players.
        
        Returns:
            Number of players
        """
        return len(self._players)
    
    def is_ready_for_game(self):
        """Check if player setup is valid for starting game.
        
        Returns:
            True if player count is within valid range
        """
        return self.MIN_PLAYERS <= len(self._players) <= self.MAX_PLAYERS
    
    def clear_players(self):
        """Clear all players (for reset/new game)."""
        self._players = []
